src/native/watchdog/watchdog.cpp

```diff
@@ -0,0 +1,139 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#include <cstdio>
+#include <cstdlib>
+#include <errno.h>
+#include <signal.h>
+
+#ifdef TARGET_WINDOWS
+
+#include <windows.h>
+#include <string>
+
+#else // !TARGET_WINDOWS
+
+#include <chrono>
+#include <sys/wait.h>
+#include <thread>
+#include <unistd.h>
+#include <vector>
+
+#endif // TARGET_WINDOWS
+
+int run_timed_process(const long, const int, const char *[]);
+
+#ifdef TARGET_X86
+int __cdecl main(const int argc, const char *argv[])
+#else
+int main(const int argc, const char *argv[])
+#endif
+{
+    if (argc < 3)
+    {
+        printf("There are missing arguments. Got %d instead of 3+ :(\n", argc);
+        return EXIT_FAILURE;
+    }
+
+    const long timeout_sec = strtol(argv[1], nullptr, 10);
+    int exit_code = run_timed_process(timeout_sec * 1000L, argc-2, &argv[2]);
+
+    printf("App Exit Code: %d\n", exit_code);
+    return exit_code;
+}
+
+int run_timed_process(const long timeout_ms, const int proc_argc, const char *proc_argv[])
+{
+#ifdef TARGET_WINDOWS
+    std::string cmdline(proc_argv[0]);
+
+    for (int i = 1; i < proc_argc; i++)
+    {
+        cmdline.append(" ");
+        cmdline.append(proc_argv[i]);
+    }
+
+    STARTUPINFOA startup_info;
+    PROCESS_INFORMATION proc_info;
+    unsigned long exit_code;
+
+    ZeroMemory(&startup_info, sizeof(startup_info));
+    startup_info.cb = sizeof(startup_info);
+    ZeroMemory(&proc_info, sizeof(proc_info));
+
+    if (!CreateProcessA(NULL, &cmdline[0], NULL, NULL, FALSE, 0, NULL, NULL,
+                       &startup_info, &proc_info))
+    {
+        int error_code = GetLastError();
+        printf("Process creation failed... Code %d.\n", error_code);
+        return error_code;
+    }
+
+    WaitForSingleObject(proc_info.hProcess, timeout_ms);
+    GetExitCodeProcess(proc_info.hProcess, &exit_code);
+
+    CloseHandle(proc_info.hProcess);
+    CloseHandle(proc_info.hThread);
+    return exit_code;
+
+#else // !TARGET_WINDOWS
+
+    // TODO: Describe what the 'ms_factor' is, and why it's being used here.
```

Comment:
Address this TODO?

Reply:
Oops forgot that. Thanks for the catch.

Reply:
I actually removed it altogether. I needed it because originally, I was dealing with some C stuff that combined microseconds and milliseconds. After switching to C++'s `std::chrono::milliseconds`, testing with and without it yielded virtually the same wait time, so it ended up being redundant.